### PR TITLE
Add more ProgressBar and related tests and fix a parameter name

### PR DESCRIPTION
**Pull request by hughbe**
## Proposed Changes
- Add ProgressBar tests
- Fix MarqueeAnimationSpeed parameter name

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2141)

### CODE REVIEW

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ControlTests.cs [hunk 1]
```diff
@@ -12393,22 +12615,196 @@ public void Control_RescaleConstantsForDpi_Invoke_Nop(int deviceDpiOld, int devi
             control.RescaleConstantsForDpi(deviceDpiOld, deviceDpiNew);
         }
 
+        [Fact]
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ControlTests.cs]:
```suggestion
        [WinFormsFact]
```

Applies to other new tests

**Comment by hughbe** [comment]
@zsd4yr any insights as to why this should hang on our test runs? I genuinely can't judge why

**Comment by RussKie** [comment]
The master appears to be in a bad shape. I'll be looking at all the deadlocks and hangs we get.

**Comment by hughbe** [comment]
@zsd4yr any idea why these tests would hang?

**Comment by hughbe** [comment]
@JeremyKuhne any idea why this hangs? Can't seem to repro

**Comment by JeremyKuhne** [comment]
> @JeremyKuhne any idea why this hangs? Can't seem to repro

Likely my bug in the `NoAssertContext`. It was fixed 3 days ago.

**Comment by hughbe** [comment]
Hopefully the rebase will fix this then!

**Comment by hughbe** [comment]
@jeremykuhne doesn't look like it sadly :(

**Comment by hughbe** [comment]
@jeremykuhne doesn't look like it sadly

**Comment by RussKie** [comment]
I'll have a look in the next few days

**Comment by RussKie** [comment]
The tests appear to cause deadlocks...

```
Not Flagged		0x00001CE4	0x00000011	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x00003270	0x0000000F	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x00006224	0x0000000E	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x000070CC	0x00000013	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x00007AD8	0x00000001	Main Thread	Main Thread	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitOneNoCheck
Not Flagged		0x000072FC	0x00000010	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x00005BA8	0x00000021	Worker Thread	<No Name>	
Not Flagged		0x00005AFC	0x00000015	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x00006D14	0x00000009	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitOneNoCheck
Not Flagged		0x000089A0	0x0000007B	Worker Thread	<No Name>	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.CreateHandle
Not Flagged		0x0000898C	0x00000060	Worker Thread	<No Name>	System.Windows.Forms.dll!System.Windows.Forms.Application.ComponentManager.Interop.Mso.IMsoComponentManager.FPushMessageLoop
Not Flagged		0x0000899C	0x0000007A	Worker Thread	<No Name>	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.CreateHandle
Not Flagged		0x00000710	0x0000000C	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x00008994	0x00000088	Worker Thread	<No Name>	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.CreateHandle
Not Flagged		0x00006D08	0x00000016	Worker Thread	.NET SystemEvents	Microsoft.Win32.SystemEvents.dll!Microsoft.Win32.SystemEvents.WindowThreadProc
Not Flagged		0x00006E4C	0x0000000D	Worker Thread	<No Name>	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.CreateHandle
Not Flagged		0x00003D78	0x00	Worker Thread	<No Name>	
Not Flagged	>	0x000089A4	0x0000007C	Worker Thread	<No Name>	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.CreateHandle
Not Flagged		0x00007068	0x00	Worker Thread	<No Name>	
Not Flagged		0x00006750	0x00000014	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x00002478	0x0000000A	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x00005E5C	0x00000002	Worker Thread	<No Name>	
Not Flagged		0x00004108	0x00000012	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x000064F4	0x0000000B	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
Not Flagged		0x000071BC	0x00	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Diagnostics.Tracing.EventProvider.EtwEnableCallBack
```

A number of threads are blocked attempting to create handles:
image
```
>	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.CreateHandle(System.Windows.Forms.CreateParams cp = {System.Windows.Forms.CreateParams}) Line 426	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.CreateHandle() Line 5040	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Application.MarshalingControl.MarshalingControl() Line 24	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Application.ThreadContext.MarshalingControl.get() Line 374	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.WindowsFormsSynchronizationContext.WindowsFormsSynchronizationContext() Line 36	C#	Symbols loaded.
 	Xunit.StaFact.dll!Xunit.Sdk.WinFormsSynchronizationContextAdapter.Create()	Unknown	Non-user code. Skipped loading symbols.
 	Xunit.StaFact.dll!Xunit.Sdk.UITestInvoker.RunAsync.AnonymousMethod__2_0()	Unknown	Non-user code. Skipped loading symbols.
 	Xunit.StaFact.dll!Xunit.Sdk.UITestInvoker.RunOnSTA.AnonymousMethod__0()	Unknown	Non-user code. Skipped loading symbols.
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown	Non-user code. Skipped loading symbols.
```

The thread that is likely to be causing the lock is this: 
```
Not Flagged	>	0x0000898C	0x00000060	Worker Thread	<No Name>	System.Windows.Forms.dll!System.Windows.Forms.Application.ComponentManager.Interop.Mso.IMsoComponentManager.FPushMessageLoop
```

```
 	[Managed to Native Transition]		Annotated Frame
 	System.Windows.Forms.dll!System.Windows.Forms.Application.ComponentManager.Interop.Mso.IMsoComponentManager.FPushMessageLoop(System.UIntPtr dwComponentID = 0x0000000000000001, Interop.Mso.msoloop uReason = ModalForm, void* pvLoopData = 0x0000000000000000) Line 396	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Application.ThreadContext.RunMessageLoopInner(Interop.Mso.msoloop reason = ModalForm, System.Windows.Forms.ApplicationContext context = {System.Windows.Forms.Application.ModalApplicationContext}) Line 1129	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Application.ThreadContext.RunMessageLoop(Interop.Mso.msoloop reason = ModalForm, System.Windows.Forms.ApplicationContext context = {System.Windows.Forms.Application.ModalApplicationContext}) Line 987	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Application.RunDialog(System.Windows.Forms.Form form = {System.Windows.Forms.ThreadExceptionDialog}) Line 1223	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Form.ShowDialog(System.Windows.Forms.IWin32Window owner = null) Line 5713	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Form.ShowDialog() Line 5599	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Application.ThreadContext.OnThreadException(System.Exception t = {"Assert.Same() Failure\r\nExpected: (null)\r\nActual:   EventArgs { }"}) Line 902	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Application.OnThreadException(System.Exception t = {"Assert.Same() Failure\r\nExpected: (null)\r\nActual:   EventArgs { }"}) Line 1109	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.WndProcException(System.Exception e = {"Assert.Same() Failure\r\nExpected: (null)\r\nActual:   EventArgs { }"}) Line 13617	C#	Symbols loaded.
>	System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnThreadException(System.Exception e = {"Assert.Same() Failure\r\nExpected: (null)\r\nActual:   EventArgs { }"}) Line 59	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.Callback(System.IntPtr hWnd = 0x0000000000132200, Interop.User32.WindowMessage msg = WM_CREATE, System.IntPtr wparam = 0x0000000000000000, System.IntPtr lparam = 0x0000007a5fffd5e0) Line 388	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
 	[Managed to Native Transition]		Annotated Frame
 	System.Windows.Forms.dll!System.Windows.Forms.NativeWindow.CreateHandle(System.Windows.Forms.CreateParams cp = {System.Windows.Forms.CreateParams}) Line 459	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.CreateHandle() Line 5040	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ProgressBar.CreateHandle() Line 564	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.Handle.get() Line 2569	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.Control.SendMessage(int msg = 0x00000406, int wparam = 0x00000000, int lparam = 0x00000064) Line 10712	C#	Symbols loaded.
 	System.Windows.Forms.dll!System.Windows.Forms.ProgressBar.OnHandleCreated(System.EventArgs e = null) Line 598	C#	Symbols loaded.
 	System.Windows.Forms.Tests.dll!System.Windows.Forms.Tests.ProgressBarTests.SubProgressBar.OnHandleCreated(System.EventArgs e = null) Line 2603	C#	Symbols loaded.
 	System.Windows.Forms.Tests.dll!System.Windows.Forms.Tests.ProgressBarTests.ProgressBar_OnHandleCreated_Invoke_CallsHandleCreated(System.Windows.Forms.ProgressBarStyle style = Blocks, System.EventArgs eventArgs = null) Line 2022	C#	Symbols loaded.
 	[Native to Managed Transition]		Annotated Frame
```

This thread is started here:
image

I'm guessing attempts to create a dialog, we lock all other threads until the modal dialog is closed.

@JeremyKuhne @weltkante thoughts? 
Am I mis-interpreting things again? 😊

**Comment by hughbe** [comment]
Something weird is happening. Basically I think a test is failing (see below). This causes an exception to be thrown and this seems to block everything. Is this a side effect of using WinFormsFact?

System.Windows.Forms.dll!System.Windows.Forms.Control.ControlNativeWindow.OnThreadException(System.Exception e = {"Assert.Same() Failure\r\nExpected: (null)\r\nActual:   EventArgs { }"}) Line 59	C#	Symbols loaded.

**Comment by hughbe** [comment]
Yeah. The problem here is that one of the tests failed and we end up calling `ThreadContext.OnThreadException`

For some reason `SystemInformation.UserInteractive` is `true` (although obviously the CI is not interactive) so we end up trying to display a dialog.

IMO this is a bug in our testing platform (and `WinFormsFact` etc.). Either we should register a  `_threadExceptionHandler` or force `SystemInformation.UserInteractive` to be `false`.

```
            /// <summary>
            ///  Called when an untrapped exception occurs in a thread.  This allows the
            ///  programmer to trap these, and, if left untrapped, throws a standard error
            ///  dialog.
            /// </summary>
            internal void OnThreadException(Exception t)
            {
                if (GetState(STATE_INTHREADEXCEPTION))
                {
                    return;
                }

                SetState(STATE_INTHREADEXCEPTION, true);
                try
                {
                    if (_threadExceptionHandler != null)
                    {
                        _threadExceptionHandler(Thread.CurrentThread, new ThreadExceptionEventArgs(t));
                    }
                    else
                    {
                        if (SystemInformation.UserInteractive)
                        {
                            ThreadExceptionDialog td = new ThreadExceptionDialog(t);
                            DialogResult result = DialogResult.OK;

                            try
                            {
                                result = td.ShowDialog();
                            }
                            finally
                            {
                                td.Dispose();
                            }
                            switch (result)
                            {
                                case DialogResult.Abort:
                                    ExitInternal();
                                    Environment.Exit(0);
                                    break;
                                case DialogResult.Yes:
                                    if (t is WarningException w)
                                    {
                                        Help.ShowHelp(null, w.HelpUrl, w.HelpTopic);
                                    }
                                    break;
                            }
                        }
                        else
                        {
                            // Ignore unhandled thread exceptions. The user can
                            // override if they really care.
                        }
                    }
                }
                finally
                {
                    SetState(STATE_INTHREADEXCEPTION, false);
                }
            }
```

**Comment by hughbe** [comment]
OK there may also be a bug in `OnHandleCreated` implementation of `ProgressBar`.

Basically, if you call `ProgressBar.OnHandleCreated` without a `Handle`, you force creation of `Handle`. I think this may cause deadlock somehow as `OnHandleCreated` would be called again once the window receives a `WM_CREATE` message

```cs
protected override void OnHandleCreated(EventArgs e)
{
    base.OnHandleCreated(e);
    SendMessage(NativeMethods.PBM_SETRANGE32, _minimum, _maximum);
    SendMessage(NativeMethods.PBM_SETSTEP, _step, 0);
    SendMessage(NativeMethods.PBM_SETPOS, _value, 0);
    UnsafeNativeMethods.SendMessage(new HandleRef(this, Handle), NativeMethods.PBM_SETBKCOLOR, 0, ColorTranslator.ToWin32(BackColor));
    UnsafeNativeMethods.SendMessage(new HandleRef(this, Handle), NativeMethods.PBM_SETBARCOLOR, 0, ColorTranslator.ToWin32(ForeColor));
    StartMarquee();
    SystemEvents.UserPreferenceChanged += new UserPreferenceChangedEventHandler(UserPreferenceChangedHandler);
}
```
```

**Comment by weltkante** [comment]
>  if you call `ProgressBar.OnHandleCreated` without a `Handle`

why would you do that? one of the assumptions of that method is that a handle has been created. I'm pretty sure no OnHandledCreated implementation checks IsHandleCreated and just assumes its true

**Comment by hughbe** [comment]
> why would you do that? one of the assumptions of that method is that a handle has been created. I'm pretty sure no OnHandledCreated implementation checks IsHandleCreated and just assumes its true

This is not the case. See for example, `Control.OnHandleCreated`

```cs
        /// <summary>
        ///  Inheriting classes should override this method to find out when the
        ///  handle has been created.
        ///  Call base.OnHandleCreated first.
        /// </summary>
        [EditorBrowsable(EditorBrowsableState.Advanced)]
        protected virtual void OnHandleCreated(EventArgs e)
        {
            if (IsHandleCreated)
            {
                // Setting fonts is for some reason incredibly expensive.
                // (Even if you exclude font handle creation)
                if (!GetStyle(ControlStyles.UserPaint))
                {
                    SetWindowFont();
                }

                if (DpiHelper.IsPerMonitorV2Awareness && !(typeof(Form).IsAssignableFrom(GetType())))
                {
                    int old = _deviceDpi;
                    _deviceDpi = (int)UnsafeNativeMethods.GetDpiForWindow(new HandleRef(this, HandleInternal));
                    if (old != _deviceDpi)
                    {
                        RescaleConstantsForDpi(old, _deviceDpi);
                    }
                }

                // Restore dragdrop status. Ole Initialize happens
                // when the ThreadContext in Application is created
                SetAcceptDrops(AllowDrop);

                Region region = (Region)Properties.GetObject(s_regionProperty);
                if (region != null)
                {
                    IntPtr regionHandle = GetHRgn(region);

                    try
                    {
                        if (IsActiveX)
                        {
                            regionHandle = ActiveXMergeRegion(regionHandle);
                        }

                        if (UnsafeNativeMethods.SetWindowRgn(new HandleRef(this, Handle), new HandleRef(this, regionHandle), SafeNativeMethods.IsWindowVisible(new HandleRef(this, Handle))) != 0)
                        {
                            //The HWnd owns the region.
                            regionHandle = IntPtr.Zero;
                        }
                    }
                    finally
                    {
                        if (regionHandle != IntPtr.Zero)
                        {
                            Gdi32.DeleteObject(regionHandle);
                        }
                    }
                }

                // Cache Handle in a local before asserting so we minimize code running under the Assert.
                IntPtr handle = Handle;

                if (Properties.GetObject(s_accessibilityProperty) is ControlAccessibleObject accObj)
                {
                    accObj.Handle = handle;
                }
                if (Properties.GetObject(s_ncAccessibilityProperty) is ControlAccessibleObject ncAccObj)
                {
                    ncAccObj.Handle = handle;
                }

                // Set the window text from the Text property.
                if (_text != null && _text.Length != 0)
                {
                    User32.SetWindowTextW(new HandleRef(this, Handle), _text);
                }

                if (!(this is ScrollableControl) && !IsMirrored && GetExtendedState(ExtendedStates.SetScrollPosition) && !GetExtendedState(ExtendedStates.HaveInvoked))
                {
                    BeginInvoke(new EventHandler(OnSetScrollPosition));
                    SetExtendedState(ExtendedStates.HaveInvoked, true);
                    SetExtendedState(ExtendedStates.SetScrollPosition, false);
                }

                // Listen to UserPreferenceChanged if the control is top level and interested in the notification.
                if (GetExtendedState(ExtendedStates.InterestedInUserPreferenceChanged))
                {
                    ListenToUserPreferenceChanged(GetTopLevel());
                }
            }
            
            ((EventHandler)Events[s_handleCreatedEvent])?.Invoke(this, e);

            if (IsHandleCreated)
            {
                // Now, repost the thread callback message if we found it.  We should do
                // this last, so we're as close to the same state as when the message
                // was placed.
                if (GetState(States.ThreadMarshalPending))
                {
                    User32.PostMessageW(this, s_threadCallbackMessage);
                    SetState(States.ThreadMarshalPending, false);
                }
            }
        }
```

I think this may be precisely due to the logic here - calling `this.Handle` in `OnHandleCreated` would end up calling `OnHandleCreated` again, perhaps causing deadlock?

**Comment by hughbe** [comment]
Woot! The tests are fixed!

**Comment by weltkante** [comment]
> See for example, Control.OnHandleCreated

Well, thats the superclass, I meant subclass implementations, sorry for not being specific. I've had a look again and most subclasses just do `SendMessage` or use the `Handle` without any check on IsHandleCreated (there are a few exceptions like CheckBox and RadioButton but the majority ignores IsHandleCreated). I'm pretty sure its not in the design of this API to be called without a handle, even though the base class protects against it, its probably just to limit damage when something goes wrong.

If you want this changed I'd suggest moving it to a separate issue so all control subclasses are updated to respect IsHandleCreated in OnHandleCreated overrides.

**Comment by hughbe** [comment]
> I'm pretty sure its not in the design of this API to be called without a handle, even though the base class protects against it, its probably just to limit damage when something goes wrong.

Right. In my opinion it is an oversight in the base classes. In terms of unblocking this PR we need this fix, however

**Comment by RussKie** [comment]
> Woot! The tests are fixed!

Yay! How did it happen though and how can we sure it won't start deadlocking all tests again? 😨

**Comment by hughbe** [comment]
> Yay! How did it happen though and how can we sure it won't start deadlocking all tests again? 😨

So basically I think there were two problems
- Firstly, if an exception is handled using `Application.OnThreadException` then we will hang 100% of the time. This should be somehow fixed in #2234 
- Secondly, there was a timing bug (depending on when the `WM_CREATE`) message is received where `OnHandleCreated` would be called twice and the `HandleCreated` event would be called twice. The test `ProgressBar_OnHandleCreated_Invoke_CallsHandleCreated` checked for the specific `EventArgs` and failed because it didn't expect the `EventArgs` to be `EventArgs.Empty` but rather `null` which was the value passed to `OnHandleCreated`. This test-failed exception occurred inside the `NativeWindow` class and was caught by a catch block resulting in a call to `Application.OnThreadException` and the resulting hang

This doesn't repro on my .NET Framework VM but that may be due to the timing of events and the configuration of that specific system.

**Comment by RussKie** [comment]
We still see random test hangs (e.g. [this](https://dev.azure.com/dnceng/public/_build/results?buildId=405449&view=logs&jobId=7c8326b9-0a5f-532a-e6de-db8515c72d9a)), so your hypothesis above sounds plausible.

Nice work lads, thank you!

**Pull request merged**